### PR TITLE
Add leak-tracked ref counting to MultiSearchResponse

### PR DESCRIPTION
**Pull request by original-brownbear**
Make the ref-counting leak tracked and release child items for MultiSearchResponse and adjust tests + REST action accordingly.

for #102030

### CODE REVIEW

**Comment by original-brownbear** [comment]
WIP for now lets see what tests say first + this has a dependency on https://github.com/elastic/elasticsearch/pull/102607 which should be discussed first

**Comment by original-brownbear** [comment]
extracted the general transport fix to https://github.com/elastic/elasticsearch/pull/102696 lets review that first maybe :)

**Review by kpollich — APPROVED** [review]
Fleet codeowner 🚀

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponse.java
Comment (original-brownbear) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponse.java:172]: No these aren't releasable as of right now. Not sure it's worth adding it either. Maybe we could add that interface to transport messages but that's a discussion for distrib and some other time I think :)

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java
Comment (benwtrent) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java:240]: This is tricky to me, `shardResponses` is a map outside the listener and keeps a copy of the `MultiSearchResponse`. This happens `itemsPerIndex` times. Each time, the listener "exits" with the map keeping a reference.

Then once all the searches are completed, it is reduced, potentially creating a NEW multisearchresponse with references to the response items in the individually collected responses.

I wonder if `shardResponses` should increment the references when storing the values, and then once this is all completed, decrement them.

Comment (original-brownbear) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java:240]: Hmm you're right there's a massive testing gap here in the enrich codebase. We only ever run with the counter going up to `1` so we just never hit this. It's definitely a bug, I'll address enhancing the tests first in a separate PR tomorrow and then we can get back to this, lets not mix this stuff :)

Comment (original-brownbear) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java:240]: https://github.com/elastic/elasticsearch/pull/102751 once that's in we should see this issue reproduce here

Comment (original-brownbear) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java:241]: Alright, I merged the above PR in here and it reproduced this issue. I fixed it just the way you describe above and it should be fine now.  If you're fine with this now and reviewing the other PR takes time, I can just back the test out of this PR before merging :)

**Review by piergm — APPROVED** [review]
LGTM!

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java
Comment (benwtrent) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java:242]:
```suggestion
                        response.incRef(); // will be released during reduce
```

**Review by benwtrent — APPROVED** [review]
I think this looks good now. Shows Enrich needs more tests!

**Comment by original-brownbear** [comment]
Thanks all!

**Pull request merged**
